### PR TITLE
Add automatic generation of CRDs for CNP and CCNP

### PR DESCRIPTION
**Pull request by christarazi**
This paves the path for generating CRDs using [controller-tools](https://github.com/kubernetes-sigs/controller-tools)'s [controller-gen tool](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/controller-gen). This will help in preparation for the Kubernetes 1.19 release. A significant benefit is the ability to generate the validation schemas, instead of writing them by hand, which can be error-prone.

See commit msgs.

Updates https://github.com/cilium/cilium/pull/11476
Related https://github.com/cilium/cilium/issues/11142

### CODE REVIEW

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
CI has passed! :tada: Now to do the clean up and resolve the leftover TODOs.

**Comment by christarazi** [comment]
At a high-level, the remaining items for this PR are as follows:

- [ ] Address all TODO items
- [ ] Address commits marked as "wip"
- [ ] Figure out what to do with the private fork of controller-tools. See 64ac4ce for more details / context.
  - [ ] File issue regarding the custom JSON marshaller bug where CRD generation is skipped because a type implements custom JSON marshaling. This happens with `api.Rule`.
  - [ ] Lack of upstream `oneOf` support
    - Open question: using the basic `oneOf` support implemented in the fork, how can we validate that CNP only allows endpointSelector, while CCNP allows both, but not at the same time?
      - One approach could be to separate out the CRD struct types, instead of how it is today where CCNP embeds CNP. This will allow us finer-grained control, but a lot of code depends on this relationship and will have to be modified. A minor issue in the grand scheme of things.
        - Add unit tests to cover this if chosen.
- [ ] Go though the diff of the previous CRD YAML vs new CRD YAML:
```
vimdiff <(yq -y '.spec.validation.openAPIV3Schema.properties.spec.properties' <(g show master:examples/crds/ciliumnetworkpolicies.yaml)) <(yq -y '.spec.validation.openAPIV3Schema.properties.spec.properties' <(tail -n +3 examples/crds/ciliumnetworkpolicies.yaml))
```

**Comment by christarazi** [comment]
test-me-please

File: pkg/azure/types/types.go
Comment (aanm) [inline pkg/azure/types/types.go]: they were not exported, so they should be `json:"-"`

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/cnp_types.go]: I think we can replace this `Timestamp` struct, i.e. remove it entirely, and replace all usages with a `github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1.Time` instead.

File: operator/k8s_node.go
Comment (aanm) [inline operator/k8s_node.go]: This could be done in a different commit, like the one where you replaced the timestamp with metav1.Time

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: Yes, it is still needed, it's used to perform the validation in the preflight command

**Review by aanm — CHANGES_REQUESTED** [review]
:100: looks really good! One other thing I noticed that is missing is the lack of the annotation field containing the version number.

File: operator/k8s_node.go
Comment (christarazi) [reply operator/k8s_node.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go]: Done

File: pkg/azure/types/types.go
Comment (christarazi) [reply pkg/azure/types/types.go]: Fixed

**Comment by christarazi** [comment]
> 100 looks really good! One other thing I noticed that is missing is the lack of the annotation field containing the version number.

Fixed.

I rebased on top of master including the K8s 1.19 changes.

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
New functionality in f538ba8. Now we validate entities at the apiserver level, instead of admitting policies with invalid entities and relying on the agent to reject them.

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
CI has passed :tada:! The `lint` job is failing which another PR has fixed already, so still good to go there. All items have been addressed in https://github.com/cilium/cilium/pull/11607#issuecomment-663815263. Marking as ready for review.

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
Rebased

File: pkg/azure/api/mock/mock.go
Comment (aanm) [inline pkg/azure/api/mock/mock.go:203]: Please split this commit in 2. One that only adds the `json:"-"` and another that renames all the fields. The field renaming commit will likely be reverted in the future once https://go-review.googlesource.com/c/tools/+/245857 is merged.

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (aanm) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go]: It's not addressed.

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/cnp_types.go]: You can remove this, there's not need to justify files IMO.

File: pkg/k8s/apis/cilium.io/v2/ccnp_types.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/ccnp_types.go]: You can remove this, there's not need to justify files IMO.

File: pkg/k8s/slim/k8s/apis/meta/v1/time.go
Comment (aanm) [inline pkg/k8s/slim/k8s/apis/meta/v1/time.go:33]: I don't think this `json:"-"` is really, is it?

File: pkg/k8s/apis/cilium.io/v2/types.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/types.go]: Inconsistent with other structs, discussed offline.

File: pkg/policy/api/http.go
Comment (aanm) [inline pkg/policy/api/http.go]: Similar to what is done with `Value string `json:"value,omitempty"`` on line 62.
```
	//
	// +kubebuilder:validation:Optional
```

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `return err`

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `return err`

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/cnp_types.go:33]: This was accidental removed 2 commits behind.

File: pkg/k8s/apis/cilium.io/v2/ccnp_types.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/ccnp_types.go]: This needs to be in a separate commit because it's introducing bugs. With master branch, the json unmarshaling is happening into the fields of the embedded CiliumNetworkPolicy structure, with this change that is no longer happening which causes the following code to fail:
```golang
// DeepEqual compares 2 CNPs while ignoring the LastAppliedConfigAnnotation and
// ignoring the Status field of the CNP.
func (in *CiliumNetworkPolicy) DeepEqual(other *CiliumNetworkPolicy) bool {
	switch {
	case (in == nil) != (other == nil):
		return false
	case (in == nil) && (other == nil):
		return true
	}

	if !(in.Name == other.Name && in.Namespace == other.Namespace) {
		return false
	}

	// Ignore v1.LastAppliedConfigAnnotation annotation
	lastAppliedCfgAnnotation1, ok1 := in.GetAnnotations()[v1.LastAppliedConfigAnnotation]
	lastAppliedCfgAnnotation2, ok2 := other.GetAnnotations()[v1.LastAppliedConfigAnnotation]
	defer func() {
		if ok1 && in.GetAnnotations() != nil {
			in.GetAnnotations()[v1.LastAppliedConfigAnnotation] = lastAppliedCfgAnnotation1
		}
		if ok2 && other.GetAnnotations() != nil {
			other.GetAnnotations()[v1.LastAppliedConfigAnnotation] = lastAppliedCfgAnnotation2
		}
	}()
	delete(in.GetAnnotations(), v1.LastAppliedConfigAnnotation)
	delete(other.GetAnnotations(), v1.LastAppliedConfigAnnotation)

	return comparator.MapStringEquals(in.GetAnnotations(), other.GetAnnotations()) &&
		in.deepEqual(other)
}
```

If we call the DeepEqual method from a CCNP structure, it will use the DeepEqual method of the CNP since it's embedded, as expected. The problem rises that the CNP's DeepEqual is accessing fields of itself, so `in.Name`, `in.GetAnnotations()`, etc, will always be empty because the unmarshalling happen into the `metav1.ObjectMeta` and `metav1.TypeMeta` of the CCNP and not the CNP like it was previously.

A suggestion to fix this is to have a DeepEqual wrapper in the CCNP which has similar code in the CNP's DeepEqual:
```
func (in *CiliumClusterwideNetworkPolicy) DeepEqual(other *CiliumClusterwideNetworkPolicy) bool {
```
(Note: It's probably better to refactor the entire code of the CNP's DeepEqual, except the in.deepEqual(other), into a dedicated function so that we cna use it in the CCNP's DeepEqual and CNP's DeepEqual.)

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `return err`

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `return err`

Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `return err`

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `s/Panic/Fatal` to crash Cilium.

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `s/Panic/Fatal` to crash Cilium.

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/client/register.go]: `s/Panic/Fatal` to crash Cilium.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/slim/k8s/apis/meta/v1/time.go
Comment (christarazi) [reply pkg/k8s/slim/k8s/apis/meta/v1/time.go:33]: `controller-gen` complains about a field missing a JSON tag if I don't have this change.

File: pkg/azure/api/mock/mock.go
Comment (christarazi) [reply pkg/azure/api/mock/mock.go:203]: Done

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go]: Removed

File: pkg/k8s/apis/cilium.io/v2/ccnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/ccnp_types.go]: Removed

File: pkg/k8s/apis/cilium.io/v2/types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/types.go]: Done

File: pkg/policy/api/http.go
Comment (christarazi) [reply pkg/policy/api/http.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go:33]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/client/register.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/client/register.go]: Fixed

File: pkg/k8s/apis/cilium.io/v2/ccnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/ccnp_types.go]: Implemented in 5432c71b8cdb022c9951e7d07606115d1e4e0402. Is this what you had in mind? Ideally this should be tested somewhere (JSON marshalling and unmarshalling), but I couldn't find a single spot for this. Where would you suggest adding a test to validate this?

**Comment by christarazi** [comment]
test-me-plesase

**Comment by christarazi** [comment]
test-me-plesase

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
retest-runtime

File: pkg/policy/api/selector.go
Comment (aanm) [inline pkg/policy/api/selector.go:56]: The addition of the JSON tag should have been done in the previous commit.

**Review by aanm — APPROVED** [review]
Does not require a second review from my side.

I've created this as follow ups https://github.com/cilium/cilium/issues/12737

**Review by aanm — CHANGES_REQUESTED** [review]
Sorry, I noticed that we aren't updating the `CustomResourceDefinitionSchemaVersion` version as we are changing the schema.

Also, before we merge this we should double check if the generated CRD is different than the current one. I've spotted a missing regex for the `matchLabels` field as well as it seems the properties of the `endpointSelector` seem to be empty?
```
    schema:
      openAPIV3Schema:
        description: CiliumNetworkPolicy is a Kubernetes third-party resource with
          an extended version of NetworkPolicy.
        properties:
          apiVersion:
            description: 'APIVersion defines the versioned schema of this representation
              of an object. Servers should convert recognized schemas to the latest
              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
            type: string
          kind:
            description: 'Kind is a string value representing the REST resource this
              object represents. Servers may infer this from the endpoint the client
              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
            type: string
          metadata:
            type: object
          spec:
            description: Spec is the desired Cilium specific rule specification.
            oneOf:
            - properties:
                endpointSelector: {}
              required:
              - endpointSelector
            - properties:
                nodeSelector: {}
              required:
              - nodeSelector
```

**Comment by christarazi** [comment]
> Sorry, I noticed that we aren't updating the `CustomResourceDefinitionSchemaVersion` version as we are changing the schema.

I'll make a commit to update the schema version.
 
> Also, before we merge this we should double check if the generated CRD is different than the current one. I've spotted a missing regex for the `matchLabels` field as well as it seems the properties of the `endpointSelector` seem to be empty?
> 

I don't see a `matchLabels` field with a regex in the original CRD YAML. All `matchLabels` fields are accounted for in the new CRD YAML. Can you point this one out?

The properties of `endpointSelector` here are empty because of the `oneOf` block. If you look further into the CRD YAML, you'll find its full definition.

**Comment by aanm** [comment]
> > Sorry, I noticed that we aren't updating the `CustomResourceDefinitionSchemaVersion` version as we are changing the schema.
> 
> I'll make a commit to update the schema version.
> 
> > Also, before we merge this we should double check if the generated CRD is different than the current one. I've spotted a missing regex for the `matchLabels` field as well as it seems the properties of the `endpointSelector` seem to be empty?
> 
> I don't see a `matchLabels` field with a regex in the original CRD YAML. All `matchLabels` fields are accounted for in the new CRD YAML. Can you point this one out?

The CRD Yaml in the repo must be outdated
```
$ kubectl get crd ciliumnetworkpolicies.cilium.io -o yaml | grep -C4 matchLabels
                            matchLabels:
                              additionalProperties:
                                maxLength: 63
                                pattern: ^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$
                                type: string
                              description: matchLabels is a map of {key,value} pairs.
                                A single {key,value} in the matchLabels map is equivalent
                                to an element of matchExpressions, whose key field
                                is "key", the operator is "In", and the values array
                                contains only "value". The requirements are ANDed.
                              type: object
```
> 
> The properties of `endpointSelector` here are empty because of the `oneOf` block. If you look further into the CRD YAML, you'll find its full definition.

**Comment by christarazi** [comment]
> The CRD Yaml in the repo must be outdated

Good catch; I've fixed it. It's part of the `Add markers for schema generation` commit. Will look into further potential diffs between the CRD registered and the new CRD YAML.

> I thought the `endpointSelector` would need to have its properties each time it shows up but since it's auto-generated it's probably ok.

:+1:

File: pkg/policy/api/selector.go
Comment (christarazi) [reply pkg/policy/api/selector.go:56]: Fixed

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
Looks like the latest failures are legit. Forcing the CRD schema validation for `matchLabels` to be `string` is applied to the type, instead of in `additionalProperties`. The complication here is that it doesn't seem like `controller-gen` supports inserting into the `additionalProperties` field. We may need to implement support for this in the fork.

**Comment by christarazi** [comment]
Needed to push https://github.com/christarazi/controller-tools/commit/1b051103a13760ad557ad7df5b2c022287ce9ff4 to our fork in order to support eb09959 for `matchLabels`. Without the commit to the fork, we would've had to define a new type (instead of a type alias) which would have required changing every usage of `matchLabels` in the codebase (a **huge** amount of code).

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
Rebased

**Comment by christarazi** [comment]
Rebased again

**Comment by christarazi** [comment]
As discussed offline, the following additions were made:

- Removed `GO111MODULE` from `make manifests` target
- Updated dev environment requirements to include go-bindata
- Consolidated `make manifest` and `make go-bindata` to run inside `./contrib/scripts/check-k8s-code-gen.sh` so that it is part of the GH Action, preventing CRD drift
  - Updated `generate-k8s-api` GH Action to install go-bindata as a prerequisite

**Comment by christarazi** [comment]
test-me-please

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/cnp_types.go]: This was addressed in a later commit which replaces the Timestamp type. This commit is just a "rename".

**Review by aanm — APPROVED** [review]
:100:

**Comment by aanm** [comment]
Travis was green in https://travis-ci.com/github/cilium/cilium/jobs/369231108

**Pull request merged**
